### PR TITLE
Cleanup some code in System.Linq.Expressions

### PR DESCRIPTION
**Pull request by jamesqo**
Removed some redundant allocations caused by creating a `List` and immediately calling `ToArray` on it. Also some code cleanup: removed some unnecessary semicolons/null checks, simplified if statements, reduced indentation.

cc @stephentoub

### CODE REVIEW

File: src/System.Linq.Expressions/src/System/Linq/Expressions/IndexExpression.cs [hunk 1]
```diff
@@ -457,51 +456,42 @@ private static void ValidateAccessor(Expression instance, MethodInfo method, Par
 
         private static void ValidateAccessorArgumentTypes(MethodInfo method, ParameterInfo[] indexes, ref ReadOnlyCollection<Expression> arguments)
         {
-            if (indexes.Length > 0)
+            if (indexes.Length != arguments.Count)
             {
-                if (indexes.Length != arguments.Count)
+                throw Error.IncorrectNumberOfMethodCallArguments(method);
+            }
+
```
Comment (JonHanna) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/IndexExpression.cs]: If we're no longer going to take the shorter path on `indexes.Length == 0` at the start, could we `return` in such a case?

Comment (jamesqo) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/IndexExpression.cs]: @JonHanna There's also an `else if` clause below that which essentially runs if both collections' counts are 0 (and has the exact same effect), so I don't think it makes a difference.

**Comment by stephentoub** [comment]
cc: @VSadov

File: src/System.Linq.Expressions/src/System/Linq/Expressions/IndexExpression.cs [hunk 2]
```diff
@@ -272,14 +272,13 @@ private static string GetArgTypesString(Expression[] arguments)
 
         private static PropertyInfo FindProperty(Type type, string propertyName, Expression[] arguments, BindingFlags flags)
         {
-            var props = type.GetProperties(flags).Where(x => x.Name.Equals(propertyName, StringComparison.CurrentCultureIgnoreCase)); ;
-            PropertyInfo[] members = new List<PropertyInfo>(props).ToArray();
-            if (members == null || members.Length == 0)
+            var members = type.GetProperties(flags)
+                .Where(x => x.Name.Equals(propertyName, StringComparison.CurrentCultureIgnoreCase));
+            if (!members.Any())
```
Comment (stephentoub) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/IndexExpression.cs]: Have you measured the perf throughput here?  Yes, your change removes the list/array allocation, which is good.  But you've also caused both the Any and the Where to allocate extra enumerators where they wouldn't have before, in order to enumerate in the Any() call, and this causes the Where to need to enumerate until either the end of the source or until the predicate returns true, whichever comes first.

Why is this !members.Any() call needed at all?  Without it, we'll simply call FindBestProperty, which will also do the enumeration (just the once), and if it's empty, FindBestProperty will return 0 and we'll end up returning null from the method anyway.

File: src/System.Linq.Expressions/src/System/Linq/Expressions/IndexExpression.cs [hunk 3]
```diff
@@ -457,51 +456,42 @@ private static void ValidateAccessor(Expression instance, MethodInfo method, Par
 
         private static void ValidateAccessorArgumentTypes(MethodInfo method, ParameterInfo[] indexes, ref ReadOnlyCollection<Expression> arguments)
         {
-            if (indexes.Length > 0)
+            if (indexes.Length != arguments.Count)
```
Comment (stephentoub) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/IndexExpression.cs]: There are no tests covering some of the paths being changed here.  Before we go trying to change code like this, at a minimum anything being changed needs test coverage.

File: src/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/InterpretedFrame.cs [hunk 4]
```diff
@@ -157,9 +157,9 @@ public IEnumerable<InterpretedFrameInfo> GetStackTraceDebugInfo()
 
         internal void SaveTraceToException(Exception exception)
         {
-            if (exception.Data[typeof(InterpretedFrameInfo)] == null)
+            if (!exception.Data.Contains(typeof(InterpretedFrameInfo)))
```
Comment (stephentoub) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/InterpretedFrame.cs]: Why did this change?

File: src/System.Linq.Expressions/src/System/Linq/Expressions/IndexExpression.cs (hunk 2 above)
Comment (jamesqo) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/IndexExpression.cs]: You're correct, I think I understand what you're saying (`GetEnumerator` simply returns the same object for iterator state machines, but needs to allocate a new one if it's called twice). I'll remove the `Any` call.

File: src/System.Linq.Expressions/src/System/Linq/Expressions/IndexExpression.cs (hunk 3 above)
Comment (jamesqo) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/IndexExpression.cs]: You're probably right; I've traced down where this method is being called from (self-note: `Expression.Property` with something returning not-null for `GetGetMethod`). Will add the corresponding tests tomorrow.

**Comment by stephentoub** [comment]
@dotnet-bot test this please

**Comment by stephentoub** [comment]
@jamesqo, are you still working on this PR, or should it be closed?  I didn't see the tests referred to added.  Thanks.

**Comment by jamesqo** [comment]
I'm going to close this for now; I had forgotten about this pull request, and the semantic changes to the code in question don't seem too important. Thanks for reminding me, @stephentoub.

**Pull request closed without merging**